packages/office-ui-fabric-react/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -142,15 +142,15 @@ export class FocusTrapZone extends BaseComponent<IFocusTrapZoneProps, {}> implem
       return;
     }
 
-    const _firstFocusableChild = getFirstFocusable(this._root.current, this._root.current.firstChild as HTMLElement, true);
-    const _lastFocusableChild = getLastTabbable(this._root.current, this._root.current.lastChild as HTMLElement, true);
+    const _firstTabbableChild = getFirstTabbable(this._root.current, this._root.current.firstChild as HTMLElement, true);
```

Comment:
This will change the behavior of FocusTrapZone so this is a breaking change... Instead consider adding a prop to use focusable vs tabbable

Reply:
I argue that this is a bug fix.  The previous behavior allows focus to leave the trap zone which goes against the entire purpose of the control.  There isn't a scenario where that behavior is desirable.

Reply:
It's not clear to me how getFirstFocusable would go outside of the trap zone? Also this could break FTZ navigation when focusable components are not contained within a focus zone (really, if they aren't tabbable), correct?

Reply:
The unit test I created is the repro.  Remove the code fix and the unit test will fail.

The scenario is where the first child in the FTZ is a FZ, and the FZ surrounds a list containing focusable items.  You focus on one of the list items that is not the first one.  Then hit shift-tab and pop out of the FTZ.  There is a related scenario involving the last child being a FZ.

FTZ works by determining the first and last things the focus can be on, and if you are on the first and press shift-tab or if you are on the last and hit tab, it wraps and jumps focus to the opposite end.  Else it does nothing.

In this example you are on the "first" thing because the FZ has marked all of the other list items it contains with tabIndex=-1 (using the roving tabindex technique) which means they are focusable but not tabbable.  But the FTZ doesn't look at tabbability right now--it ignores the tabIndex--so it finds the first focusable thing as the first item in the list.  This is not the same as where you currently are so it thinks there is nothing to do and ignores the shift-tab keypress.  The browser takes over and pops you out of the FTZ.

You can see a live repro by going to https://wacappribbons.azurewebsites.net, click Word, click Changes in the top-right,  click Get Wywa, click an item in the list (not the first), hit shift-tab.

As to breaking FTZ navigation:  This change is not predicated on having a FZ involved.  The search will still scan for the first focusable thing with tabIndex!=-1 (i.e. tabbable) and find components in or out of FZs.  There are other existing unit tests are cover these non-FZ cases that are still passing.

Reply:
I think that could be an issue with the creator having component3 be focused and having tabindex=-1. What's happening is the FocusTrapZone is seeing that it's not at either end of itself and it is determining to let the default browser behavior move focus, but since everything is tabindex=-1 focus moves out of the FocusTrapZone.

We need to make sure we aren't breaking the FocusTrapZone for people who are using it who might not be setting tabindex (not every component or set of components uses roving tab index). Does getFirstTabbable (or getLastTabbable) return an element as tabbable if there is no tabindex? If so, we may be fine.

I just want to make sure that if the FocusTrapZone gets to the end and wants to put focus back on the first item, it should be able to do that even if the tabindex=-1. 

Consider the case where there is a FocusTrapZone with only a FocusZone inside, if the user is focused on the last item in the FocusZone and then hits tab (and the FocusZone is not set to circularNavigate or it is set to not handle TAB keys) then the FocusTrapZone should move focus to the first item which I believe will be broken if your fix goes in (I think it will not move focus with your fix, which is not correct).

Reply:
getFirstTabbable/getLastTabbable call isElementTabbable(elem, checkTabIndex:true) which will return true for elements without a tabIndex set.

I disagree with the final statement.  tabIndex=-1 doesn't mean it can't be focused, just that the tab key should never take you there.  That means that in the case of the FTZ with a single FZ inside, tab should not move you to the top item in the FZ.  The FZ is like a giant tab-stop with sub-navigation via arrow keys.  Why would surrounding it with an FTZ change those semantics?

To be clear, that example was for the **existing** code.  Any concerns there are around use of tabbable are already present in the existing code because of the call to getLastTabbable, and will cause a problem in the opposite direction (tab vs. shift-tab).  Consider the case of a single FZ in an FTZ with three items in the FZ, and the middle item is focused.  Pressing tab and shift-tab both do the same thing: they jump from the middle item to the first item, which is broken.  Further, pressing tab or shift-tab again both do the same thing: nothing.

Reply:
We shouldn't alter the existing behavior unless the new behavior encompasses the existing behavior, which I'm pretty sure is not the case currently. We don't want to potentially regress people who are using it